File: src/mcp/connector_tools.zig
//! MCP `connector_test` tool: exercise a connector through its deterministic
//! local path. Extracted from handlers.zig — these routines hold no global
//! state, so they take only an allocator plus request fields.

const std = @import("std");
const abi = @import("abi");

/// Run the deterministic local path for one connector and return a one-line
/// status string. Unknown services return `error.UnknownConnector`.
pub fn runConnectorTest(allocator: std.mem.Allocator, service: []const u8, input: []const u8) ![]u8 {
    const connectors = abi.connectors;

    if (std.mem.eql(u8, service, "openai")) {
        var client = connectors.openai.Client.init(allocator, .{ .api_key = "mcp-local-key", .base_url = "https://api.openai.com" });
        defer client.deinit();
        const messages = try buildUserMessages(allocator, input);
        defer allocator.free(messages);
        var response = try client.chatCompletion(allocator, "gpt-local", messages);
        defer response.deinit(allocator);
        return try std.fmt.allocPrint(allocator, "connector=openai status={d} body={s}", .{ response.status, response.body });
    }

    if (std.mem.eql(u8, service, "anthropic")) {
        var client = connectors.anthropic.Client.init(allocator, .{ .api_key = "mcp-local-key", .base_url = "https://api.anthropic.com" });
        defer client.deinit();
        var response = try client.message(allocator, "claude-local", input, 256);
        defer response.deinit(allocator);
        return try std.fmt.allocPrint(allocator, "connector=anthropic status={d} body={s}", .{ response.status, response.body });
    }

    if (std.mem.eql(u8, service, "discord")) {
        var bot = connectors.discord.Bot.init(allocator, .{ .token = "mcp-local-token", .client_id = "123456789012345678" });
        defer bot.deinit();
        try bot.connect();
        const body = try bot.sendMessage(allocator, "234567890123456789", input);
        defer allocator.free(body);
        return try std.fmt.allocPrint(allocator, "connector=discord status=200 body={s}", .{body});
    }

    if (std.mem.eql(u8, service, "twilio")) {
        var client = connectors.twilio.Client.init(allocator, connectors.twilio.TwilioConfig.local());
        defer client.deinit();
        var response = try client.handleConversationRelayEvent(allocator, .{
            .kind = .user_transcript,
            .conversation_id = "CA" ++ "0123456789abcdef0123456789abcdef",
            .customer_id = "+15551234567",
            .transcript = input,
        }, "ABI local relay acknowledged.");
        defer response.deinit(allocator);
        return try std.fmt.allocPrint(allocator, "connector=twilio status=200 body={s}", .{response.text});
    }

    if (std.mem.eql(u8, service, "grok")) {
        var client = connectors.grok.Client.init(allocator, connectors.grok.grokConfig());
        defer client.deinit();
        const messages = try buildUserMessages(allocator, input);
        defer allocator.free(messages);
        var response = try client.chatCompletion(allocator, "grok-local", messages);
        defer response.deinit(allocator);
        return try std.fmt.allocPrint(allocator, "connector=grok status={d} body={s}", .{ response.status, response.body });
    }

    return error.UnknownConnector;
}

fn buildUserMessages(allocator: std.mem.Allocator, input: []const u8) ![]u8 {
    var out: std.ArrayListUnmanaged(u8) = .empty;
    errdefer out.deinit(allocator);
    try out.appendSlice(allocator, "[{\"role\":\"user\",\"content\":");
    try abi.connectors.json.appendJsonString(&out, allocator, input);
    try out.appendSlice(allocator, "}]");
    return try out.toOwnedSlice(allocator);
}

test {
    std.testing.refAllDecls(@This());
}

File: crates/abi-mcp/src/connector_tools.rs
//! MCP `connector_test` tool: exercise a connector through its deterministic
//! local path.
//!
//! Ported from `src/mcp/connector_tools.zig`, reusing the already-ported
//! `abi-connectors` clients (including Twilio `ConversationRelay` local synthesis).

use abi_connectors::{
    Client, ConnectorConfig, ConversationRelayEvent, build_local_conversation_response, payload,
};

/// Why [`run_connector_test`] could not produce a response.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ConnectorToolError {
    /// The connector's local synthesis failed unexpectedly.
    Failed,
}

/// Run the deterministic local path for one connector and return a one-line
/// status string, matching `connector_tools.runConnectorTest` in Zig.
///
/// `service` is assumed already validated against the tool's enum (see
/// `handlers::CONNECTOR_TEST_FIELDS`) — this only distinguishes ported vs.
/// unported services among the five valid ones.
pub fn run_connector_test(service: &str, input: &str) -> Result<String, ConnectorToolError> {
    match service {
        "openai" => {
            let messages = build_user_messages(input);
            let client = Client::openai(ConnectorConfig::new(
                "mcp-local-key",
                "https://api.openai.com",
            ));
            let response = client
                .complete("gpt-local", &messages)
                .map_err(|_| ConnectorToolError::Failed)?;
            Ok(format!(
                "connector=openai status={} body={}",
                response.status, response.body
            ))
        }
        "anthropic" => {
            // The MCP tool asks for max_tokens=256, not abi-connectors'
            // general-purpose default of 4096 — call the payload builders
            // directly rather than through `Client::complete`.
            const MAX_TOKENS: u32 = 256;
            let body = payload::build_anthropic_body("claude-local", input, MAX_TOKENS, false);
            let text =
                payload::anthropic_local_response("claude-local", input, MAX_TOKENS, body.len());
            Ok(format!("connector=anthropic status=200 body={text}"))
        }
        "discord" => {
            let body = payload::discord_local_ack("234567890123456789", input);
            Ok(format!("connector=discord status=200 body={body}"))
        }
        "grok" => {
            let messages = build_user_messages(input);
            let client = Client::grok(ConnectorConfig::new("mcp-local-key", "https://api.x.ai"));
            let response = client
                .complete("grok-local", &messages)
                .map_err(|_| ConnectorToolError::Failed)?;
            Ok(format!(
                "connector=grok status={} body={}",
                response.status, response.body
            ))
        }
        "twilio" => {
            // MCP fixture: fixed conversation/customer ids + local agent reply,
            // matching Zig `handleConversationRelayEvent(.user_transcript, …)`.
            let event = ConversationRelayEvent::user_transcript(input);
            let response =
                build_local_conversation_response(&event, "ABI local relay acknowledged.");
            Ok(format!(
                "connector=twilio status=200 body={}",
                response.text
            ))
        }
        _ => Err(ConnectorToolError::Failed),
    }
}

fn build_user_messages(input: &str) -> String {
    let mut out = String::from(r#"[{"role":"user","content":"#);
    out.push_str(&abi_foundation::json::escape_json_string(input));
    out.push_str("}]");
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn openai_matches_the_golden_local_synthesis() {
        let text = run_connector_test("openai", "ping").expect("openai is ported");
        assert_eq!(
            text,
            "connector=openai status=200 body={\"choices\":[{\"message\":{\"role\":\"assistant\",\"content\":\"OpenAI-compatible local response model=gpt-local messages_bytes=34 request_bytes=67\"}}]}"
        );
    }

    #[test]
    fn twilio_local_relay_synthesizes_body() {
        // "hi" is < 3 chars and escalates as low_confidence (Zig parity).
        let text = run_connector_test("twilio", "what are your hours?").expect("twilio is ported");
        assert_eq!(
            text,
            "connector=twilio status=200 body=ABI local relay acknowledged."
        );
        let escalate =
            run_connector_test("twilio", "let me talk to a human").expect("twilio escalate");
        assert!(
            escalate.contains("support specialist"),
            "expected escalation reply: {escalate}"
        );
        let short = run_connector_test("twilio", "hi").expect("short escalates");
        assert!(
            short.contains("support specialist"),
            "short transcript should escalate: {short}"
        );
    }

    #[test]
    fn anthropic_and_discord_and_grok_synthesize_locally() {
        assert!(
            run_connector_test("anthropic", "hi")
                .expect("anthropic is ported")
                .starts_with("connector=anthropic status=200 body=")
        );
        assert!(
            run_connector_test("discord", "hi")
                .expect("discord is ported")
                .starts_with("connector=discord status=200 body=")
        );
        assert!(
            run_connector_test("grok", "hi")
                .expect("grok is ported")
                .starts_with("connector=grok status=200 body=")
        );
    }
}
